### ISSUE TITLE
return a bookmark event with bookmarkafterresourceversion immediately to reduce WatchList time cost(1~1.25s)

### ISSUE THREAD

**Issue by likakuli** [kubernetes/kubernetes#122277]
### What happened?

the resource version of the bookmark event returned by watchlist maybe larger than bookmarkafterresourceversion. which means there need more 1 ~ 1.25s to return even the event with bookmarkresourceversion already received in cacher.

### What did you expect to happen?

return a bookmark event with bookmarkafterresourceversion immediately when to sendWatchCacheEvent in cacheWatcher if the resource version already not elder than bookmarkafterresourceversion.
DON‘T wait another cycle of bookmark timer.

### How can we reproduce it (as minimally and precisely as possible)?

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L899
there is a bookmark timer triggered every 1 ~ 1.25s.
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/cacher/cache_watcher.go#L527-L529
```golang
if event.ResourceVersion > resourceVersion || (event.Type == watch.Bookmark && event.ResourceVersion == resourceVersion && !c.wasBookmarkAfterRvSent()) {
    c.sendWatchCacheEvent(event)
}
```
the resourceversion passed to process function is just bookmarkafterresourversion.
consider the following scenario

1. bookmark event come at t1 with resourceversion < bookmarkresourceversion, no action
2. normal event come at t2 with resourceversion = bookmarkresourceversion, no action
3. normal event come at t3 with resourceversion > bookmarkresourceversion, then the event will be send to client
4. bookmark event come at t4 with resourceversion > bookmarkresourceversion, then the bookmark will be set with an special annotation and return to client

t4 - t1 = 1s ~ 1.25s

the final bookmark event can be send back to client at t1. this will reduce one bookmark timer cycle for watchlist request in some scenario. In extreme cases, this time factor may lead to a timeout for the watchlist request.

### Kubernetes version

<details>

```console
$ kubectl version
# master branch
```

</details>

### Cloud provider

<details>
none
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by likakuli** [comment]
@wojtek-t @p0lyn0mial can you please help to review is this expected

**Comment by Anuragt1104** [comment]
Hey, I can work on this issue. Can you assign me?

**Comment by wojtek-t** [comment]
I'm not yet convinced we should work on that.
As soon as we have RequestProgressNotify:
https://github.com/kubernetes/kubernetes/blob/d19420f3ec9d360dbc1bc57044182a993b2a958a/staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress.go#L99
feature working, we will bring this down to 100ms or so, which no longer justifies this change [which will introduce another complexity to already highly complex code].

**Comment by wojtek-t** [comment]
@p0lyn0mial - FYI

**Comment by likakuli** [comment]
> we will bring this down to 100ms or so, which no longer justifies this change [which will introduce another complexity to already highly complex code].

make sense. 

there is another case which i don't known if it's expected. For client not use any sdk just call api directly, if they want to change List to WatchList, the result returned by WatchList have all possible event type and they need to process these events by event type just like what client-go do. If there treat all event as `ADD`, the WatchList result maybe wrong for them because there will be some event with same object(ns/name) but with different event type returned by WatchList.

**Comment by wojtek-t** [comment]
> there is another case which i don't known if it's expected. For client not use any sdk just call api directly, if they want to change List to WatchList, the result returned by WatchList have all possible event type and they need to process these events by event type just like what client-go do. If there treat all event as ADD, the WatchList result maybe wrong for them because there will be some event with same object(ns/name) but with different event type returned by WatchList.

I don't expect random clients to use WatchList API directly.

I was chatting with @p0lyn0mial earlier today about switching things like List methods of our client libraries (or kubectl in particular) to use WatchList underneath. But for direct API users I don't expect them to use WatchList unless they know how to handle it. So that is "working as intended"

**Comment by p0lyn0mial** [comment]
> I'm not yet convinced we should work on that.
As soon as we have RequestProgressNotify:

[kubernetes/staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress.go](https://github.com/kubernetes/kubernetes/blob/d19420f3ec9d360dbc1bc57044182a993b2a958a/staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress.go#L99)
> err := pr.requestWatchProgress(ctx) 

> feature working, we will bring this down to 100ms or so, which no longer justifies this change [which will introduce another complexity to already highly complex code].

We still might need to do some changes to the cacher because the bookmark events are delivered at most every ~1 second.
So, it looks like simply calling `requestWatchProgress` won't be enough if we would like to decrease the latency to 100ms.

Regarding the second question, I agree with @wojtek-t . The WatchList is a different kind of API, not a drop-in replacement for the List API.

**Comment by wojtek-t** [comment]
> We still might need to do some changes to the cacher because the bookmark events are delivered at most every ~1 second.
So, it looks like simply calling requestWatchProgress won't be enough if we would like to decrease the latency to 100ms.

That's not fully true - the whole idea of requestWatchProgress is that it delivers progress-notify on-demand and we can use that to deliver bookmark within 100ms. We may want some super minor adjustments to process them correctly in watchcache [can't remember the code from the top of my head], but that [if even needed], will be extremely minor change.

**Comment by likakuli** [comment]
> we will bring this down to 100ms or so, which no longer justifies this change [which will introduce another complexity to already highly complex code].

checked the current logic, the `requestWatchProgress` will be called only when to call List function. if there is no running List request, there will be no requestWatchProgress function call. thus `WatchList` also have 1s~1.25s latency.

**Comment by likakuli** [comment]
also i think there may be a potential risks about the current logic of `requestWatchProgress `, i will open a new issue to describe it.

**Comment by likakuli** [comment]
after digging, seems there is a `defaultBookmarkFrequency` variable set to 1m, so even bookmark timer is triggered every 1s ~ 1.25s, but the latency will be up to 1m? 

please let me if i'm wrong @wojtek-t @p0lyn0mial

**Comment by p0lyn0mial** [comment]
> after digging, seems there is a defaultBookmarkFrequency variable set to 1m, so even bookmark timer is triggered every 1s ~ 1.25s, but the latency will be up to 1m?
>
> please let me if i'm wrong @wojtek-t @p0lyn0mial

The bookmark delivery time can be configured (`a`, `b` and `c` from the link below). By default, it is set to 1 minute. Have a look at https://github.com/kubernetes/kubernetes/blob/9ce03d9254cbd47d82a4b2d88e2a915943d90e26/staging/src/k8s.io/apiserver/pkg/storage/cacher/cache_watcher.go#L222

In general, there will be some changes to the code required if we want to decrease the delivery time to 100ms.

**Comment by likakuli** [comment]
@wojtek-t @p0lyn0mial I have a question here. why use global RV as the bookmarkRV? is it possible to use the max RV of the objects? 
e.g. there is a watch request with rv unset to get all pods in default namespace with label app=test by WatchList
global RV=101 and max RV of pod in default namespace is 90
for such watch request, is it ok to use 90 as the bookmarkafterRV to reduce the latency caused by watch progress request interval(5s).

**Comment by p0lyn0mial** [comment]
> @wojtek-t @p0lyn0mial I have a question here. why use global RV as the bookmarkRV? is it possible to use the max RV of the related objects?

the global rv helps ensure that returned data remains consistent, as if it were served from etcd. Without the RV we have no idea how far behind etcd we are and we are vulnerable to stale reads (like [vulnerable](https://github.com/kubernetes/kubernetes/issues/59848))

**Comment by likakuli** [comment]
i think use max RV for a key or key prefix from etcd also can reach this goal.

**Comment by p0lyn0mial** [comment]
> i think use max RV for a key or key prefix from etcd also can reach this goal.

Are you sure there is a maximum RV for a key or prefix in etcd? If I recall correctly, etcd maintains a single global RV, not per key or prefix.

**Comment by likakuli** [comment]
> If I recall correctly, etcd maintains a single global RV, not per key or prefix.

yes, etcd use a single global RV, but we can use `clientv3.WithLastRev()` to get the latest mod RV for the specified resource. e.g. /pods, /services

**Comment by seans3** [comment]
/assign @serathius 

Is this an issue in an area you're working on? What do you think?

/triage accepted

**Comment by serathius** [comment]
I don't have enough context on WatchList, assigning to feature owner.
/unassign
/assign @p0lyn0mial
